### PR TITLE
MULE-11941: Operation, Scope and Router models should extend ProcessorModel

### PR DESCRIPTION
**Pull request by alegmarra**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+import org.mule.runtime.extension.api.error.MuleErrors;
+
+import com.google.gson.reflect.TypeToken;
+
+/**
+ * An {@link ExtensionDeclarer} for Mule's Core Runtime
+ *
+ * @since 4.0
+ */
+class MuleExtensionModelDeclarer {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java]: este declaration si se puede ver!

File: core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+        .withParent(newError("ANY", "MULE").build()).build());
+  }
+
+  private void declareAsync(ExtensionDeclarer extensionDeclarer, ClassTypeLoader typeLoader) {
+    ConstructDeclarer async = extensionDeclarer.withConstruct("async")
+        .describedAs("Processes the nested list of message processors asynchronously using a thread pool");
+
+    async.withChain();
+    async.onDefaultParameterGroup()
+        .withOptionalParameter("schedulerName") //TODO "name" => "schedulerName"
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java]: @pablolagreca rename of scheduler

File: core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+        .describedAs(
+                     "Allows a \u0027flow\u0027 to be referenced such that the message processing will continue in the referenced flow "
+                         + "before returning. Message processing in the referenced \u0027flow\u0027 will occur within the context of the "
+                         + "referenced flow and will therefore use its exception strategy etc.");
+
+    flowRef.withOutput().ofType(BaseTypeBuilder.create(JAVA).anyType().build());
+    flowRef.withOutputAttributes().ofType(BaseTypeBuilder.create(JAVA).anyType().build());
+
+    flowRef.onDefaultParameterGroup()
+        .withRequiredParameter("flowName") //TODO rename "name" => "flowName"
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java]: @pablolagreca rename flowRef name

File: core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+
+    ConstructDeclarer forEach = extensionDeclarer.withConstruct("foreach")
+        .describedAs(
+                     "The foreach Processor allows iterating over a collection payload, or any collection obtained by an expression,"
+                         + " generating a message for each element.");
+
+    forEach.withChain();
+
+    forEach.onDefaultParameterGroup()
+        .withOptionalParameter("elements") //TODO rename "collection" => "elements"
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java]: @pablolagreca rename foreach's `collection` to a plural name

File: core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+    //TODO errorHandler.isOneRouteRequired(true);
+
+  }
+
+  private void declareOnErrorRoute(ClassTypeLoader typeLoader, NestedRouteDeclarer onError) {
+    onError.withChain();
+
+    onError.onDefaultParameterGroup()
+        .withOptionalParameter("handlerName") //TODO name => handlerName
```
Comment (alegmarra) [inline core/src/main/java/org/mule/runtime/core/api/extension/MuleExtensionModelDeclarer.java]: @pablolagreca rename of on-error elements name to handlerName

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java [hunk 6]
```diff
@@ -188,7 +163,7 @@ public void load() throws Exception {
   public static ExtensionModel loadExtension(Class<?> clazz, ExtensionModelLoader loader) {
     Map<String, Object> params = new HashMap<>();
     params.put(TYPE_PROPERTY_NAME, clazz.getName());
-    params.put(VERSION, MULE_VERSION);
+    params.put(VERSION, "4.0.0-SNAPSHOT");
```
Comment (alegmarra) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java]: hack for intellij but also OK ¯\_(ツ)_/¯

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java [hunk 7]
```diff
@@ -539,7 +539,7 @@ public static ExtensionModel loadExtension(Class<?> clazz) {
 
   public static ExtensionModel loadExtension(Class<?> clazz, Map<String, Object> params) {
     params.put(TYPE_PROPERTY_NAME, clazz.getName());
-    params.put(VERSION, getProductVersion());
+    params.put(VERSION, "4.0.0-SNAPSHOT");
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/MuleExtensionUtils.java]: same hackButAlsoOk

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java (hunk 6 above)
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/resources/ExtensionModelJsonGeneratorTestCase.java]: not ok. This is going to break the moment we branch for GA. Fix it in another PR though

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterModelValidatorTestCase.java [hunk 8]
```diff
@@ -104,12 +103,4 @@ public void invalidModelDueToDefaultValueWhenRequired() {
     validate(extensionModel, validator);
   }
 
-  @Test(expected = IllegalModelDefinitionException.class)
-  public void invalidModelDueToListWithoutPluralName() {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterModelValidatorTestCase.java]: this is wrong. You only made the plural rule specific to the java loader. That doesn't mean you shouldn't have test for it. 

Also, where's the SourceDeclarationEnricher you removed in the ext-api PR?

Comment (alegmarra) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterModelValidatorTestCase.java]: SourceDeclarationEnricher  => RedeliveryPolicyDeclarationEnricher + ReconnectionStrategyDeclarationEnricher

**Pull request merged**